README.rst
Very simple Python library for color and formatting in terminal.
Collection of color codes and names for 256 color terminal setups.
The following is a list of 256 colors for Xterm, containing an example
of the displayed color, Xterm Name, Xterm Number and HEX.

.. image:: https://gitlab.com/dslackw/images/raw/master/colored/logo.png
    :target: https://gitlab.com/dslackw/colored

`[CHANGELOG] <https://gitlab.com/dslackw/colored/blob/master/CHANGELOG>`_

Video Demo
----------

.. image:: https://gitlab.com/dslackw/images/raw/master/colored/video.png
    :target: https://asciinema.org/a/10712

The following colors works with most terminals and terminals emulators.
ANSI/VT100 escape sequences can be used in every programming languages.

Attributes:

.. code-block:: bash

    +-----+------------------+
    |Code | Description      |
    +-----+------------------+
    |  1  | bold             |
    |  2  | dim              |
    |  4  | underlined       |
    |  5  | blink            |
    |  7  | reverse          |
    |  8  | hidden           |
    |  0  | reset            |
    |  21 | res_bold         |
    |  22 | res_dim          |
    |  24 | res_underlined   |
    |  25 | res_blink        |
    |  27 | res_reverse      |
    |  28 | res_hidden       |
    +------------------------+


256 Foreground and Background Colors - Full Chart:

.. code-block:: bash

    +-----+---------------------+
    |Code | Description         |
    +-----+---------------------+
    | 0   | black               |
    | 1   | red                 |
    | 2   | green               |
    | 3   | yellow              |
    | 4   | blue                |
    | 5   | magenta             |
    | 6   | cyan                |
    | 7   | light_gray          |
    | 8   | dark_gray           |
    | 9   | light_red           |
    | 10  | light_green         |
    | 11  | light_yellow        |
    | 12  | light_blue          |
    | 13  | light_magenta       |
    | 14  | light_cyan          |
    | 15  | white               |
    | 16  | grey_0              |
    | 17  | navy_blue           |
    | 18  | dark_blue           |
    | 19  | blue_3a             |
    | 20  | blue_3b             |
    | 21  | blue_1              |
    | 22  | dark_green          |
    | 23  | deep_sky_blue_4a    |
    | 24  | deep_sky_blue_4b    |
    | 25  | deep_sky_blue_4c    |
    | 26  | dodger_blue_3       |
    | 27  | dodger_blue_2       |
    | 28  | green_4             |
    | 29  | spring_green_4      |
    | 30  | turquoise_4         |
    | 31  | deep_sky_blue_3a    |
    | 32  | deep_sky_blue_3b    |
    | 33  | dodger_blue_1       |
    | 34  | green_3a            |
    | 35  | spring_green_3a     |
    | 36  | dark_cyan           |
    | 37  | light_sea_green     |
    | 38  | deep_sky_blue_2     |
    | 39  | deep_sky_blue_1     |
    | 40  | green_3b            |
    | 41  | spring_green_3b     |
    | 42  | spring_green_2a     |
    | 43  | cyan_3              |
    | 44  | dark_turquoise      |
    | 45  | turquoise_2         |
    | 46  | green_1             |
    | 47  | spring_green_2b     |
    | 48  | spring_green_1      |
    | 49  | medium_spring_green |
    | 50  | cyan_2              |
    | 51  | cyan_1              |
    | 52  | dark_red_1          |
    | 53  | deep_pink_4a        |
    | 54  | purple_4a           |
    | 55  | purple_4b           |
    | 56  | purple_3            |
    | 57  | blue_violet         |
    | 58  | orange_4a           |
    | 59  | grey_37             |
    | 60  | medium_purple_4     |
    | 61  | slate_blue_3a       |
    | 62  | slate_blue_3b       |
    | 63  | royal_blue_1        |
    | 64  | chartreuse_4        |
    | 65  | dark_sea_green_4a   |
    | 66  | pale_turquoise_4    |
    | 67  | steel_blue          |
    | 68  | steel_blue_3        |
    | 69  | cornflower_blue     |
    | 70  | chartreuse_3a       |
    | 71  | dark_sea_green_4b   |
    | 72  | cadet_blue_2        |
    | 73  | cadet_blue_1        |
    | 74  | sky_blue_3          |
    | 75  | steel_blue_1a       |
    | 76  | chartreuse_3b       |
    | 77  | pale_green_3a       |
    | 78  | sea_green_3         |
    | 79  | aquamarine_3        |
    | 80  | medium_turquoise    |
    | 81  | steel_blue_1b       |
    | 82  | chartreuse_2a       |
    | 83  | sea_green_2         |
    | 84  | sea_green_1a        |
    | 85  | sea_green_1b        |
    | 86  | aquamarine_1a       |
    | 87  | dark_slate_gray_2   |
    | 88  | dark_red_2          |
    | 89  | deep_pink_4b        |
    | 90  | dark_magenta_1      |
    | 91  | dark_magenta_2      |
    | 92  | dark_violet_1a      |
    | 93  | purple_1a           |
    | 94  | orange_4b           |
    | 95  | light_pink_4        |
    | 96  | plum_4              |
    | 97  | medium_purple_3a    |
    | 98  | medium_purple_3b    |
    | 99  | slate_blue_1        |
    | 100 | yellow_4a           |
    | 101 | wheat_4             |
    | 102 | grey_53             |
    | 103 | light_slate_grey    |
    | 104 | medium_purple       |
    | 105 | light_slate_blue    |
    | 106 | yellow_4b           |
    | 107 | dark_olive_green_3a |
    | 108 | dark_green_sea      |
    | 109 | light_sky_blue_3a   |
    | 110 | light_sky_blue_3b   |
    | 111 | sky_blue_2          |
    | 112 | chartreuse_2b       |
    | 113 | dark_olive_green_3b |
    | 114 | pale_green_3b       |
    | 115 | dark_sea_green_3a   |
    | 116 | dark_slate_gray_3   |
    | 117 | sky_blue_1          |
    | 118 | chartreuse_1        |
    | 119 | light_green_2       |
    | 120 | light_green_3       |
    | 121 | pale_green_1a       |
    | 122 | aquamarine_1b       |
    | 123 | dark_slate_gray_1   |
    | 124 | red_3a              |
    | 125 | deep_pink_4c        |
    | 126 | medium_violet_red   |
    | 127 | magenta_3a          |
    | 128 | dark_violet_1b      |
    | 129 | purple_1b           |
    | 130 | dark_orange_3a      |
    | 131 | indian_red_1a       |
    | 132 | hot_pink_3a         |
    | 133 | medium_orchid_3     |
    | 134 | medium_orchid       |
    | 135 | medium_purple_2a    |
    | 136 | dark_goldenrod      |
    | 137 | light_salmon_3a     |
    | 138 | rosy_brown          |
    | 139 | grey_63             |
    | 140 | medium_purple_2b    |
    | 141 | medium_purple_1     |
    | 142 | gold_3a             |
    | 143 | dark_khaki          |
    | 144 | navajo_white_3      |
    | 145 | grey_69             |
    | 146 | light_steel_blue_3  |
    | 147 | light_steel_blue    |
    | 148 | yellow_3a           |
    | 149 | dark_olive_green_3  |
    | 150 | dark_sea_green_3b   |
    | 151 | dark_sea_green_2    |
    | 152 | light_cyan_3        |
    | 153 | light_sky_blue_1    |
    | 154 | green_yellow        |
    | 155 | dark_olive_green_2  |
    | 156 | pale_green_1b       |
    | 157 | dark_sea_green_5b   |
    | 158 | dark_sea_green_5a   |
    | 159 | pale_turquoise_1    |
    | 160 | red_3b              |
    | 161 | deep_pink_3a        |
    | 162 | deep_pink_3b        |
    | 163 | magenta_3b          |
    | 164 | magenta_3c          |
    | 165 | magenta_2a          |
    | 166 | dark_orange_3b      |
    | 167 | indian_red_1b       |
    | 168 | hot_pink_3b         |
    | 169 | hot_pink_2          |
    | 170 | orchid              |
    | 171 | medium_orchid_1a    |
    | 172 | orange_3            |
    | 173 | light_salmon_3b     |
    | 174 | light_pink_3        |
    | 175 | pink_3              |
    | 176 | plum_3              |
    | 177 | violet              |
    | 178 | gold_3b             |
    | 179 | light_goldenrod_3   |
    | 180 | tan                 |
    | 181 | misty_rose_3        |
    | 182 | thistle_3           |
    | 183 | plum_2              |
    | 184 | yellow_3b           |
    | 185 | khaki_3             |
    | 186 | light_goldenrod_2a  |
    | 187 | light_yellow_3      |
    | 188 | grey_84             |
    | 189 | light_steel_blue_1  |
    | 190 | yellow_2            |
    | 191 | dark_olive_green_1a |
    | 192 | dark_olive_green_1b |
    | 193 | dark_sea_green_1    |
    | 194 | honeydew_2          |
    | 195 | light_cyan_1        |
    | 196 | red_1               |
    | 197 | deep_pink_2         |
    | 198 | deep_pink_1a        |
    | 199 | deep_pink_1b        |
    | 200 | magenta_2b          |
    | 201 | magenta_1           |
    | 202 | orange_red_1        |
    | 203 | indian_red_1c       |
    | 204 | indian_red_1d       |
    | 205 | hot_pink_1a         |
    | 206 | hot_pink_1b         |
    | 207 | medium_orchid_1b    |
    | 208 | dark_orange         |
    | 209 | salmon_1            |
    | 210 | light_coral         |
    | 211 | pale_violet_red_1   |
    | 212 | orchid_2            |
    | 213 | orchid_1            |
    | 214 | orange_1            |
    | 215 | sandy_brown         |
    | 216 | light_salmon_1      |
    | 217 | light_pink_1        |
    | 218 | pink_1              |
    | 219 | plum_1              |
    | 220 | gold_1              |
    | 221 | light_goldenrod_2b  |
    | 222 | light_goldenrod_2c  |
    | 223 | navajo_white_1      |
    | 224 | misty_rose1         |
    | 225 | thistle_1           |
    | 226 | yellow_1            |
    | 227 | light_goldenrod_1   |
    | 228 | khaki_1             |
    | 229 | wheat_1             |
    | 230 | cornsilk_1          |
    | 231 | grey_100            |
    | 232 | grey_3              |
    | 233 | grey_7              |
    | 234 | grey_11             |
    | 235 | grey_15             |
    | 236 | grey_19             |
    | 237 | grey_23             |
    | 238 | grey_27             |
    | 239 | grey_30             |
    | 240 | grey_35             |
    | 241 | grey_39             |
    | 242 | grey_42             |
    | 243 | grey_46             |
    | 244 | grey_50             |
    | 245 | grey_54             |
    | 246 | grey_58             |
    | 247 | grey_62             |
    | 248 | grey_66             |
    | 249 | grey_70             |
    | 250 | grey_74             |
    | 251 | grey_78             |
    | 252 | grey_82             |
    | 253 | grey_85             |
    | 254 | grey_89             |
    | 255 | grey_93             |
    | 256 | default             |
    +-----+---------------------+

256 Colors Foreground (text):

.. image:: https://gitlab.com/dslackw/images/raw/master/colored/256_colors_fg.png
    :alt: 256 fg colors

256 Colors Background:

.. image:: https://gitlab.com/dslackw/images/raw/master/colored/256_colors_bg.png
    :alt: 256 bg colors


Installation
------------

.. code-block:: bash

    $ pip install colored --upgrade

    uninstall

    $ pip uninstall colored


Dependencies
------------

None, only Python progmamming language.

Usage Examples
--------------

How to use the module in your own python code:

.. code-block:: bash

    >>> from colored import fg, bg, attr
    >>>
    >>> print ('%s Hello World !!! %s' % (fg(1), attr(0)))
     Hello World !!!
    >>>
    >>> print ('%s%s Hello World !!! %s' % (fg(1), bg(15), attr(0)))
     Hello World !!!

Use description:

.. code-block:: bash

    >>> print ('%s%s Hello World !!! %s' % (fg('white'), bg('yellow'), attr('reset')))
     Hello World !!!
    >>>
    >>> print ('%s%s Hello World !!! %s' % (fg('orchid'), attr('bold'), attr('reset')))
     Hello World !!!
    >>>
    >>> color = bg('indian_red_1a') + fg('white')
    >>> reset = attr('reset')
    >>> print (color + 'Hello World !!!' + reset)
    Hello World !!!

Or use HEX code:

.. code-block:: bash

    >>> color = fg('#C0C0C0') + bg('#00005f')
    >>> res = attr('reset')
    >>> print (color + "Hello World !!!" + res)
    Hello World !!!

Or the convenient `stylize(text, *styles)` wrapper to save some keystrokes:

.. code-block:: bash

    >>> import colored
    >>> from colored import stylize
    >>> print(stylize("This is green.", colored.fg("green")))
    This is green.
    >>> print("This is not.")
    This is not.
    >>> angry = colored.fg("red") + colored.attr("bold")
    >>> print(stylize("This is angry text.", angry))
    This is angry text.
    >>> print(stylize("This is VERY angry text.", angry, colored.attr("underlined")))
    This is VERY angry text.
    >>> print("But this is not.")
    But this is not.

Or the variant `stylize_interactive(text, *styles)` for readline-friendliness:

.. code-block:: bash

    >>> import colored, sys
    >>> from colored import stylize_interactive, fg
    >>> sys.ps1 = stylize_interactive("myPrompt: ", fg('red'))
    myPrompt:

Use directly like `colorama <https://pypi.python.org/pypi/colorama>`_ but with more colors:

.. code-block:: bash

    >>> from colored import fore, back, style
    >>>
    >>> print (fore.LIGHT_BLUE + back.RED + style.BOLD + "Hello World !!!" + style.RESET)

Import colored module:

.. code-block:: bash

   >>> import colored
   >>>
   >>> colored.fg(1)
   '\x1b[38;5;1m'
   >>>
   >>> colored.fg(257)
   Traceback (most recent call last):
     File "<input>", line 1, in <module>
     File "/usr/lib64/python2.7/site-packages/colored/colored.py", line 381, in fg
       return colored(color).foreground()
     File "/usr/lib64/python2.7/site-packages/colored/colored.py", line 350, in foreground
       color = self.reserve_paint[str(self.color)]
   KeyError: '257'
   >>>
   >>> colored.bg(30)
   '\x1b[48;5;30m'
   >>>
   >>> colored.fore.BLUE
   '\x1b[38;5;4m'

    etc.


Screenshot:

.. image:: https://gitlab.com/dslackw/images/raw/master/colored/screenshot-2.png
    :alt: example


CHANGELOG
26-06-2018
Version 1.3.9
[Fixed] - pip install #12 #12

26-06-2018
Version 1.3.8
[Fixed] - pip install #12

24-06-2018
Version 1.3.7
[Fixed] - Fix license in setup file

24-06-2018
Version 1.3.6
[Updated] - Switch to GitLab repository
          - Switch to MIT licence

07-05-2017
Version 1.3.5
[Added] - readline-safe stylize_interactive() #6

20-01-2017
Version 1.3.4
[Fixed] - Color capitalization

[Updated] - Year license

19-11-2016
Version 1.3.3
[Updated] - Updated to make HEX codes case-insensitive #2

[Added] -  Added a `stylize()` convenience method #3 

02-11-2016
Version 1.3.2

[Updated] - Merge back and fore colors names

20-09-2016
Version 1.3.1

[Added] - Color RED_3B

[Fixed] - Python3 print tests compatibility.

20-09-2016
Version 1.3.0

[Updated] - Switch libraries to vars() method.

18-05-2016
Version 1.2.2

[Updated] - Files licence.
          - Color name dark_sea_sreen_X in dark_sea_green_X. Thanks to Clifford Ireland for 
contributing.

06-06-2015
Version 1.2.1

[Updated] - Fix Python3 install.
          - Added license each file.

06-06-2015
Version 1.2.0

[Updated] - Restore KeyError message for catching exceptions.
          - Fix foreground double key.
          - Update modules import.
          - Update tests.

25-02-2015
Version 1.1.5

[Updated] - Fix pep8 style

01-08-2014
Version 1.1.4

[Updated] -  Add egg-info directory

11-07-2014
Version 1.1.3

[Updated] - Fix version, remove attr.py conflict with colored.py

11-07-2014
Version 1.1.2

[Feature] - Added class fore, back, style

10-07-2014
Version 1.1.1

[Feature] - Added HEX code colors
[Feature] - Added test_2.py
[Feature] - Added HEX list

09-07-2014
Version 1.1.0

[Feature] - Added new test_2.py and colors_list.txt
[Updated] - Fix Python 3 support

09-07-2014
Version 1.0.9

[Updated] - Remove numbers from dictionary
            Fix KeyError

08-07-2014
Version 1.0.8

[Updated] - create alias fg(), bg(), attr()

08-07-2014
Version 1.0.7

[Updated] - Include two dictionary in one

07-07-2014
Version 1.0.6

[Updated] - Include 16 colors in 256 colors. Create class 
            and attributes.

05-07-2014
Version 1.0.5

[Feature] - Added test_3.py

04-07-2014
Version 1.0.4

[Feature] - Added test_2.py

03-07-2014
Version 1.0.3

[Updated] - Remove screenshots directory

02-07-2014
Version 1.0.2

[Feature] - Added support 256 colors

02-07-2014
Version 1.0.1

[Feature] - Added test

02-07-2014
Version 1.0.0

Released version 1.0.0 developmnet status Aplha



colored/colored.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import platform

from .hex import HEX
import sys


TTY_AWARE = False
IS_TTY = sys.stdout.isatty() and sys.stderr.isatty()

_win_vterm_mode = None

class colored(object):

    ESC = "\x1b["

    END = "m"

    PAINT = {
        "black": "0",
        "red": "1",
        "green": "2",
        "yellow": "3",
        "blue": "4",
        "magenta": "5",
        "cyan": "6",
        "light_gray": "7",
        "dark_gray": "8",
        "light_red": "9",
        "light_green": "10",
        "light_yellow": "11",
        "light_blue": "12",
        "light_magenta": "13",
        "light_cyan": "14",
        "white": "15",
        "grey_0": "16",
        "navy_blue": "17",
        "dark_blue": "18",
        "blue_3a": "19",
        "blue_3b": "20",
        "blue_1": "21",
        "dark_green": "22",
        "deep_sky_blue_4a": "23",
        "deep_sky_blue_4b": "24",
        "deep_sky_blue_4c": "25",
        "dodger_blue_3": "26",
        "dodger_blue_2": "27",
        "green_4": "28",
        "spring_green_4": "29",
        "turquoise_4": "30",
        "deep_sky_blue_3a": "31",
        "deep_sky_blue_3b": "32",
        "dodger_blue_1": "33",
        "green_3a": "34",
        "spring_green_3a": "35",
        "dark_cyan": "36",
        "light_sea_green": "37",
        "deep_sky_blue_2": "38",
        "deep_sky_blue_1": "39",
        "green_3b": "40",
        "spring_green_3b": "41",
        "spring_green_2a": "42",
        "cyan_3": "43",
        "dark_turquoise": "44",
        "turquoise_2": "45",
        "green_1": "46",
        "spring_green_2b": "47",
        "spring_green_1": "48",
        "medium_spring_green": "49",
        "cyan_2": "50",
        "cyan_1": "51",
        "dark_red_1": "52",
        "deep_pink_4a": "53",
        "purple_4a": "54",
        "purple_4b": "55",
        "purple_3": "56",
        "blue_violet": "57",
        "orange_4a": "58",
        "grey_37": "59",
        "medium_purple_4": "60",
        "slate_blue_3a": "61",
        "slate_blue_3b": "62",
        "royal_blue_1": "63",
        "chartreuse_4": "64",
        "dark_sea_green_4a": "65",
        "pale_turquoise_4": "66",
        "steel_blue": "67",
        "steel_blue_3": "68",
        "cornflower_blue": "69",
        "chartreuse_3a": "70",
        "dark_sea_green_4b": "71",
        "cadet_blue_2": "72",
        "cadet_blue_1": "73",
        "sky_blue_3": "74",
        "steel_blue_1a": "75",
        "chartreuse_3b": "76",
        "pale_green_3a": "77",
        "sea_green_3": "78",
        "aquamarine_3": "79",
        "medium_turquoise": "80",
        "steel_blue_1b": "81",
        "chartreuse_2a": "82",
        "sea_green_2": "83",
        "sea_green_1a": "84",
        "sea_green_1b": "85",
        "aquamarine_1a": "86",
        "dark_slate_gray_2": "87",
        "dark_red_2": "88",
        "deep_pink_4b": "89",
        "dark_magenta_1": "90",
        "dark_magenta_2": "91",
        "dark_violet_1a": "92",
        "purple_1a": "93",
        "orange_4b": "94",
        "light_pink_4": "95",
        "plum_4": "96",
        "medium_purple_3a": "97",
        "medium_purple_3b": "98",
        "slate_blue_1": "99",
        "yellow_4a": "100",
        "wheat_4": "101",
        "grey_53": "102",
        "light_slate_grey": "103",
        "medium_purple": "104",
        "light_slate_blue": "105",
        "yellow_4b": "106",
        "dark_olive_green_3a": "107",
        "dark_green_sea": "108",
        "light_sky_blue_3a": "109",
        "light_sky_blue_3b": "110",
        "sky_blue_2": "111",
        "chartreuse_2b": "112",
        "dark_olive_green_3b": "113",
        "pale_green_3b": "114",
        "dark_sea_green_3a": "115",
        "dark_slate_gray_3": "116",
        "sky_blue_1": "117",
        "chartreuse_1": "118",
        "light_green_2": "119",
        "light_green_3": "120",
        "pale_green_1a": "121",
        "aquamarine_1b": "122",
        "dark_slate_gray_1": "123",
        "red_3a": "124",
        "deep_pink_4c": "125",
        "medium_violet_red": "126",
        "magenta_3a": "127",
        "dark_violet_1b": "128",
        "purple_1b": "129",
        "dark_orange_3a": "130",
        "indian_red_1a": "131",
        "hot_pink_3a": "132",
        "medium_orchid_3": "133",
        "medium_orchid": "134",
        "medium_purple_2a": "135",
        "dark_goldenrod": "136",
        "light_salmon_3a": "137",
        "rosy_brown": "138",
        "grey_63": "139",
        "medium_purple_2b": "140",
        "medium_purple_1": "141",
        "gold_3a": "142",
        "dark_khaki": "143",
        "navajo_white_3": "144",
        "grey_69": "145",
        "light_steel_blue_3": "146",
        "light_steel_blue": "147",
        "yellow_3a": "148",
        "dark_olive_green_3": "149",
        "dark_sea_green_3b": "150",
        "dark_sea_green_2": "151",
        "light_cyan_3": "152",
        "light_sky_blue_1": "153",
        "green_yellow": "154",
        "dark_olive_green_2": "155",
        "pale_green_1b": "156",
        "dark_sea_green_5b": "157",
        "dark_sea_green_5a": "158",
        "pale_turquoise_1": "159",
        "red_3b": "160",
        "deep_pink_3a": "161",
        "deep_pink_3b": "162",
        "magenta_3b": "163",
        "magenta_3c": "164",
        "magenta_2a": "165",
        "dark_orange_3b": "166",
        "indian_red_1b": "167",
        "hot_pink_3b": "168",
        "hot_pink_2": "169",
        "orchid": "170",
        "medium_orchid_1a": "171",
        "orange_3": "172",
        "light_salmon_3b": "173",
        "light_pink_3": "174",
        "pink_3": "175",
        "plum_3": "176",
        "violet": "177",
        "gold_3b": "178",
        "light_goldenrod_3": "179",
        "tan": "180",
        "misty_rose_3": "181",
        "thistle_3": "182",
        "plum_2": "183",
        "yellow_3b": "184",
        "khaki_3": "185",
        "light_goldenrod_2a": "186",
        "light_yellow_3": "187",
        "grey_84": "188",
        "light_steel_blue_1": "189",
        "yellow_2": "190",
        "dark_olive_green_1a": "191",
        "dark_olive_green_1b": "192",
        "dark_sea_green_1": "193",
        "honeydew_2": "194",
        "light_cyan_1": "195",
        "red_1": "196",
        "deep_pink_2": "197",
        "deep_pink_1a": "198",
        "deep_pink_1b": "199",
        "magenta_2b": "200",
        "magenta_1": "201",
        "orange_red_1": "202",
        "indian_red_1c": "203",
        "indian_red_1d": "204",
        "hot_pink_1a": "205",
        "hot_pink_1b": "206",
        "medium_orchid_1b": "207",
        "dark_orange": "208",
        "salmon_1": "209",
        "light_coral": "210",
        "pale_violet_red_1": "211",
        "orchid_2": "212",
        "orchid_1": "213",
        "orange_1": "214",
        "sandy_brown": "215",
        "light_salmon_1": "216",
        "light_pink_1": "217",
        "pink_1": "218",
        "plum_1": "219",
        "gold_1": "220",
        "light_goldenrod_2b": "221",
        "light_goldenrod_2c": "222",
        "navajo_white_1": "223",
        "misty_rose1": "224",
        "thistle_1": "225",
        "yellow_1": "226",
        "light_goldenrod_1": "227",
        "khaki_1": "228",
        "wheat_1": "229",
        "cornsilk_1": "230",
        "grey_100": "231",
        "grey_3": "232",
        "grey_7": "233",
        "grey_11": "234",
        "grey_15": "235",
        "grey_19": "236",
        "grey_23": "237",
        "grey_27": "238",
        "grey_30": "239",
        "grey_35": "240",
        "grey_39": "241",
        "grey_42": "242",
        "grey_46": "243",
        "grey_50": "244",
        "grey_54": "245",
        "grey_58": "246",
        "grey_62": "247",
        "grey_66": "248",
        "grey_70": "249",
        "grey_74": "250",
        "grey_78": "251",
        "grey_82": "252",
        "grey_85": "253",
        "grey_89": "254",
        "grey_93": "255",
    }

    ATTRIBUTE = {
        "bold": ESC + "1" + END,
        1: ESC + "1" + END,
        "dim": ESC + "2" + END,
        2: ESC + "2" + END,
        "italic": ESC + "3" + END,
        3: ESC + "3" + END,
        "underlined": ESC + "4" + END,
        4: ESC + "4" + END,
        "blink": ESC + "5" + END,
        5: ESC + "5" + END,
        "reverse": ESC + "7" + END,
        7: ESC + "7" + END,
        "hidden": ESC + "8" + END,
        8: ESC + "8" + END,
        "reset": ESC + "0" + END,
        0: ESC + "0" + END,
        "res_bold": ESC + "21" + END,
        21: ESC + "21" + END,
        "res_dim": ESC + "22" + END,
        22: ESC + "22" + END,
        "res_italic": ESC + "23" + END,
        23: ESC + "23" + END,
        "res_underlined": ESC + "24" + END,
        24: ESC + "24" + END,
        "res_blink": ESC + "25" + END,
        25: ESC + "25" + END,
        "res_reverse": ESC + "27" + END,
        27: ESC + "27" + END,
        "res_hidden": ESC + "28" + END,
        28: ESC + "28" + END,
    }

    def __init__(self, color):
        self.color = color
        self.enable_windows_terminal_mode()
        if str(color).startswith("#"):
            self.HEX = HEX(color.lower())
        else:
            self.HEX = ""

    def attribute(self):
        """Set or reset attributes"""
        if TTY_AWARE and not IS_TTY:
            return ""
        return colored.ATTRIBUTE[self.color]

    def foreground(self):
        """Print 256 foreground colors"""
        if TTY_AWARE and not IS_TTY:
            return ""
        code = colored.ESC + "38;5;"
        if str(self.color).isdigit():
            self.reverse_dict()
            color = self.reserve_paint[str(self.color)]
            return code + colored.PAINT[color] + colored.END
        elif self.color.startswith("#"):
            return code + str(self.HEX) + colored.END
        else:
            return code + colored.PAINT[self.color] + colored.END

    def background(self):
        """Print 256 background colors"""
        if TTY_AWARE and not IS_TTY:
            return ""
        code = colored.ESC + "48;5;"
        if str(self.color).isdigit():
            self.reverse_dict()
            color = self.reserve_paint[str(self.color)]
            return code + colored.PAINT[color] + colored.END
        elif self.color.startswith("#"):
            return code + str(self.HEX) + colored.END
        else:
            return code + colored.PAINT[self.color] + colored.END

    def reverse_dict(self):
        """reverse dictionary"""
        self.reserve_paint = dict(zip(colored.PAINT.values(), colored.PAINT.keys()))

    def enable_windows_terminal_mode(self):
        '''Enable virtual terminal processing in windows terminal. Does
        nothing if not on Windows. This is based on the rejected
        enhancement <https://bugs.python.org/issue29059>.'''
        global _win_vterm_mode
        if _win_vterm_mode != None:
            return _win_vterm_mode

        # Note: Cygwin should return something like "CYGWIN_NT..."
        _win_vterm_mode = platform.system().lower() == 'windows'
        if _win_vterm_mode == False:
            return

        from ctypes import windll, c_int, byref, c_void_p
        ENABLE_VIRTUAL_TERMINAL_PROCESSING = 0x0004
        INVALID_HANDLE_VALUE = c_void_p(-1).value
        STD_OUTPUT_HANDLE = c_int(-11)

        hStdout = windll.kernel32.GetStdHandle(STD_OUTPUT_HANDLE)
        if hStdout == INVALID_HANDLE_VALUE:
            _win_vterm_mode = False
            return

        mode = c_int(0)
        ok = windll.kernel32.GetConsoleMode(c_int(hStdout), byref(mode))
        if not ok:
            _win_vterm_mode = False
            return

        mode = c_int(mode.value | ENABLE_VIRTUAL_TERMINAL_PROCESSING)
        ok = windll.kernel32.SetConsoleMode(c_int(hStdout), mode)
        if not ok:
            # Something went wrong, proably an too old version
            # that doesn't support the VT100 mode.
            # To be more certain we could check kernel32.GetLastError
            # for STATUS_INVALID_PARAMETER, but since we only enable
            # one flag we can be certain enough.
            _win_vterm_mode = False
            return

def attr(color):
    """alias for colored().attribute()"""
    return colored(color).attribute()


def fg(color):
    """alias for colored().foreground()"""
    return colored(color).foreground()


def bg(color):
    """alias for colored().background()"""
    return colored(color).background()


def stylize(text, styles, reset=True):
    """conveniently styles your text as and resets ANSI codes at its end."""
    terminator = attr("reset") if reset else ""
    return "{}{}{}".format("".join(styles), text, terminator)


def _c0wrap(styles):
    """wrap a set of ANSI styles in C0 control codes for readline safety."""
    C0_SOH = '\x01'   # mark the beginning of nonprinting characters
    C0_STX = '\x02'   # mark the end of nonprinting characters
    return "{}{}{}".format(C0_SOH, "".join(styles), C0_STX)


def stylize_interactive(text, styles, reset=True):
    """stylize() variant that adds C0 control codes (SOH/STX) for readline
    safety."""
    # problem: readline includes bare ANSI codes in width calculations.
    # solution: wrap nonprinting codes in SOH/STX when necessary.
    # see: https://gitlab.com/dslackw/colored/issues/5
    terminator = _c0wrap(attr("reset")) if reset else ""
    return "{}{}{}".format(_c0wrap(styles), text, terminator)

def set_tty_aware():
    """Makes all interactions here tty aware.  This means that if either
    stdout or stderr are directed to something other than a tty,
    colorization will not be added."""
    global TTY_AWARE
    TTY_AWARE = True
